theano/gof/cmodule.py

```diff
@@ -2310,5 +2319,314 @@ def print_command_line_error():
             return dlimport(lib_filename)
 
 
-def icc_module_compile_str(*args):
-    raise NotImplementedError()
+class ICC_compiler(Compiler):
+    # The equivalent flags of --march=native used by icpc.
+    march_flags = None
+
+    # TODO: Can't check AMD CPU
+    supports_amdlibm = True
+
+    @staticmethod
+    def version_str():
+        icc_ver = [int(n) for n in icc_version_str.split('.')[:2]]
+        if bool(icc_ver < [16, 0]):
+            _logger.warn(
+                "WARNING: Your Intel Compiler Version is %s. "
+                "For better performance, we suggest you use the latest version. " % icc_version_str)
+            _logger.info(
+                "WARNING: Your Intel Compiler Version is %s. "
+                "For better performance, we suggest you use the latest version. " % icc_version_str)
+        return theano.config.cxx + " " + icc_version_str
+
+    @staticmethod
+    def compile_args(march_flags=True):
+        cxxflags = [flag for flag in config.icc.cxxflags.split(' ') if flag]
+
+        # Add the equivalent of -xHOST flag.  We can't use
+        # -xHOST as when the compiledir is shared by multiple
+        # computers (for example, if the home directory is on NFS), this
+        # won't be optimum or cause crash depending if the file is compiled
+        # on an older or more recent computer.
+        detect_march = ICC_compiler.march_flags is None and march_flags
+        if detect_march:
+            for f in cxxflags:
+                # If the user give an -march=X parameter, don't add one ourself
+                if ((f.startswith("--march=") or
+                     f.startswith("-march=") or
+                     f.startswith("-m") or
+                     f.startswith("-x") or
+                     f.startswith("-ax")
+                 )):
+                    _logger.warn(
+                        "WARNING: your Theano flags `icc.cxxflags` specify"
+                        " an arch optimizing flags.\n"
+                        "         It is better to let Theano/icpc find it"
+                        " automatically, but we don't do it now")
+                    detect_march = False
+                    ICC_compiler.march_flags = []
+                    break
+
+        # TODO
+        if ('icpc' not in theano.config.cxx):
+            _logger.warn(
+                "OPTIMIZATION WARNING: your Theano flag `cxx` seems not to be"
+                " the icpc compiler. So we disable the compiler optimization"
+                " specific to icpc that tell to compile for a specific CPU."
+                " At worst, this could cause slow down.\n"
+                "         You can add those parameters to the compiler yourself"
+                " via the Theano flag `gcc.cxxflags`."
+            )
```

Comment:
I think we should never call this method if cxx don't have icpc. If you agree, change this to an exception.